### PR TITLE
Boundary clock testing with rapid host based L2 discovery

### PR DESCRIPTION
**Pull request by edcdavid**
This is a work in progress PR adding BC testing functionality. Please check updated conformance test documentation within this PR for details (https://github.com/openshift/ptp-operator/blob/8acd8c7e5c274ccd9986f47031e0d67cae49d4f6/test/README.md ). 
The L2 discovery feature is implemented in the following project:
https://github.com/test-network-function/l2discovery

### CODE REVIEW

File: test/conformance/test_suite_test.go [hunk 1]
```diff
@@ -4,21 +4,21 @@
 package test_test
 
 import (
-	"context"
+	//"context"
```
Comment (josephdrichard) [inline test/conformance/test_suite_test.go:7]: Is this no longer necessary?  If so, delete, don't comment out.

**Review by josephdrichard — CHANGES_REQUESTED** [review]
Can you split this PR into three separate commits (still one PR)?  One with just doc and image changes, one with the vendor changes, and one with the code changes.

File: test/conformance/test_suite_test.go (hunk 1 above)
Comment (edcdavid) [reply test/conformance/test_suite_test.go]: fixed

File: test/README.md [hunk 2]
```diff
@@ -13,16 +14,98 @@ So for instance to run in discovery mode the command line could look like this:
...
+ oc label  node <nodename>    ptp/test-bc-master=""
+```
+In addition to the label, a valid user provided ptpconfig must be present corresponding to the labelled node
+
+example BC ptp config:
+```yaml
+apiVersion: ptp.openshift.io/v1
+kind: PtpConfig
+metadata:
+  creationTimestamp: "2022-06-13T22:27:54Z"
```
Comment (aneeshkp) [inline test/README.md:34]: Remove this

File: test/README.md [hunk 3]
```diff
@@ -13,16 +14,98 @@ So for instance to run in discovery mode the command line could look like this:
...
+  generation: 1
```
Comment (aneeshkp) [inline test/README.md:35]: remove this

File: test/README.md [hunk 4]
```diff
@@ -13,16 +14,98 @@ So for instance to run in discovery mode the command line could look like this:
...
+      f:spec:
+        .: {}
+        f:profile: {}
+        f:recommend: {}
+    manager: __debug_bin
+    operation: Update
+    time: "2022-06-13T22:27:54Z"
+  name: test-bc-master
+  namespace: openshift-ptp
+  resourceVersion: "7593685"
```
Comment (aneeshkp) [inline test/README.md:49]: remove this

File: test/README.md [hunk 5]
```diff
@@ -13,16 +14,98 @@ So for instance to run in discovery mode the command line could look like this:
...
+  uid: 7e884ecd-fbd3-46d0-bb47-81504ada2bf6
+spec:
+  profile:
+  - name: test-bc-master
+    phc2sysOpts: -a -r
```
Comment (aneeshkp) [inline test/README.md:54]: remove this

File: test/README.md [hunk 6]
```diff
@@ -13,16 +14,98 @@ So for instance to run in discovery mode the command line could look like this:
...
+    ptp4lConf: |-
+      [global]
+      ptp_dst_mac 01:1B:19:00:00:00
+      p2p_dst_mac 01:80:C2:00:00:0E
+      network_transport UDPv4
```
Comment (aneeshkp) [inline test/README.md:59]: We support only L2

File: test/README.md [hunk 7]
```diff
@@ -34,4 +117,78 @@ The Test mode discovery workflow is as follows:
  - **Desired config**: the desired mode is selected at this point. The choices are Discovery, OC, BC and DualNICBC 
  - **PTP configuration**: if OC, BC or DualNICBC modes are selected, a valid default configuration is configured automatically. The output of the configuration is 1 or 2 ptpconfig objects in the openshift-ptp. If Discovery mode is selected, this configuration step is skipped.
```
Comment (aneeshkp) [inline test/README.md:118]: Does this work for DUAL NIC BC ?

File: test/utils/clean/clean.go [hunk 8]
```diff
@@ -3,39 +3,50 @@ package clean
...
-	for _, node := range nodeList.Items {
-		delete(node.Labels, utils.PtpGrandmasterNodeLabel)
-		_, err = client.Client.CoreV1().Nodes().Update(context.Background(), &node, metav1.UpdateOptions{})
+	err = DeleteLabel(utils.PtpSlaveNodeLabel)
+	if err != nil {
+		return fmt.Errorf("clean.All: fail to delete label: %s, err: %s", utils.PtpSlaveNodeLabel, err)
 	}
-
-	nodeList, err = client.Client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=", utils.PtpSlaveNodeLabel)})
+	err = DeleteLabel(utils.PtpBCMasterNodeLabel)
```
Comment (aneeshkp) [inline test/utils/clean/clean.go:43]: Isn't the PTP Master and GM  labels mean same ?

File: test/utils/clean/clean.go [hunk 9]
```diff
@@ -47,22 +58,10 @@ func Configs() error {
...
-		if ptpConfig.Name == utils.PtpGrandMasterPolicyName || ptpConfig.Name == utils.PtpSlavePolicyName {
-			err = client.Client.PtpConfigs(utils.PtpLinuxDaemonNamespace).Delete(context.Background(), ptpConfig.Name, metav1.DeleteOptions{})
-			if err != nil {
-				return fmt.Errorf("clean.All: Failed to delete ptp config %s %v", ptpConfig.Name, err)
-			}
-		}
-	}
-	for i := 0; i < 20; i++ {
-		ptpconfigList, err = client.Client.PtpConfigs(utils.PtpLinuxDaemonNamespace).List(context.Background(), metav1.ListOptions{})
+		err = client.Client.PtpConfigs(utils.PtpLinuxDaemonNamespace).Delete(context.Background(), ptpConfig.Name, metav1.DeleteOptions{})
```
Comment (aneeshkp) [inline test/utils/clean/clean.go:61]: The clean up should only clean up policy created by the test . Not the discovered (configured) PTP config

File: test/utils/consts.go [hunk 10]
```diff
@@ -13,14 +12,17 @@ const (
...
+	PtpSlaveNodeLabel               = "ptp/test-slave"
+	PtpBCMasterNodeLabel            = "ptp/test-bc-master"
+	PtpBCSlaveNodeLabel             = "ptp/test-bc-slave"
 	PtpResourcesGroupVersionPrefix  = "ptp.openshift.io/v"
 	PtpResourcesNameOperatorConfigs = "ptpoperatorconfigs"
 	NodePtpDeviceAPIPath            = "/apis/ptp.openshift.io/v1/namespaces/openshift-ptp/nodeptpdevices/"
 	ConfigPtpAPIPath                = "/apis/ptp.openshift.io/v1/namespaces/openshift-ptp/ptpconfigs"
 	PtpContainerName                = "linuxptp-daemon-container"
 	EventProxyContainerName         = "cloud-event-proxy"
+	NamespaceTesting                = "ptp-testing"
```
Comment (aneeshkp) [inline test/utils/consts.go:25]: why do we need ptp-testing ? everything should be within openshift-ptp right ?

File: test/utils/consts.go [hunk 11]
```diff
@@ -37,4 +39,6 @@ const (
 var (
 	PtpGrandMasterPolicyName = "test-grandmaster"
 	PtpSlavePolicyName       = "test-slave"
+	PtpBcMasterPolicyName    = "test-bc-master"
```
Comment (aneeshkp) [inline test/utils/consts.go:42]: Couldn't we use old node labels for this ?

File: test/README.md (hunk 7 above)
Comment (edcdavid) [reply test/README.md:118]: Dual NIC BC is not fully implemented yet, I will update the readme

File: test/utils/clean/clean.go (hunk 8 above)
Comment (edcdavid) [reply test/utils/clean/clean.go:43]: This label is to identify the boundary clock, the other new label is for boundary clock slave

File: test/utils/consts.go (hunk 11 above)
Comment (edcdavid) [reply test/utils/consts.go:42]: I'd prefer to use a clearer label for for the specific role. Makes it easier to understand and configure

File: test/README.md (hunk 2 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/README.md (hunk 3 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/README.md (hunk 4 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/README.md (hunk 5 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/README.md (hunk 6 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/README.md (hunk 7 above)
Comment (edcdavid) [reply test/README.md]: fixed

File: test/utils/consts.go (hunk 10 above)
Comment (edcdavid) [reply test/utils/consts.go]: fixed

File: test/utils/clean/clean.go (hunk 9 above)
Comment (edcdavid) [reply test/utils/clean/clean.go]: fixed

**Review by aneeshkp — APPROVED** [review]

File: test/utils/clean/clean.go [hunk 12]
```diff
@@ -3,39 +3,50 @@ package clean
...
 	"fmt"
-	"time"
 
 	"github.com/openshift/ptp-operator/test/utils"
 	"github.com/openshift/ptp-operator/test/utils/client"
+	"github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+func DeleteLabel(label string) error {
```
Comment (aneeshkp) [inline test/utils/clean/clean.go:13]: // Add comment  for exported function

File: test/utils/nodes/nodes.go [hunk 13]
```diff
@@ -40,6 +41,9 @@ func PtpEnabled(client *client.ClientSet) ([]*NodeTopology, error) {
 	nodeTopologyList := []*NodeTopology{}
 
 	nodesList, err := MatchingOptionalSelectorPTP(nodeDevicesList.Items)
+	if err != nil {
+		logrus.Errorf("error getting node list, err= %s", err)
```
Comment (aneeshkp) [inline test/utils/nodes/nodes.go:45]: if error getting nodeList, you need to return here.

File: test/utils/clean/clean.go (hunk 12 above)
Comment (edcdavid) [reply test/utils/clean/clean.go:14]: fixed

File: test/utils/clean/clean.go [hunk 14]
```diff
@@ -47,22 +58,15 @@ func Configs() error {
...
 
 	for _, ptpConfig := range ptpconfigList.Items {
-		if ptpConfig.Name == utils.PtpGrandMasterPolicyName || ptpConfig.Name == utils.PtpSlavePolicyName {
+		if ptpConfig.Name == utils.PtpGrandMasterPolicyName ||
+			ptpConfig.Name == utils.PtpSlavePolicyName ||
+			ptpConfig.Name == utils.PtpBcMasterPolicyName ||
+			ptpConfig.Name == utils.PtpBcSlavePolicyName {
 			err = client.Client.PtpConfigs(utils.PtpLinuxDaemonNamespace).Delete(context.Background(), ptpConfig.Name, metav1.DeleteOptions{})
 			if err != nil {
 				return fmt.Errorf("clean.All: Failed to delete ptp config %s %v", ptpConfig.Name, err)
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply test/utils/clean/clean.go]: fixed

File: test/utils/daemonsets/daemonsets.go [hunk 15]
```diff
@@ -140,7 +140,7 @@ func DeleteDaemonSet(daemonSetName, namespace string) error {
 func doesDaemonSetExist(daemonSetName, namespace string) bool {
 	_, err := client.Client.DaemonSets(namespace).Get(context.TODO(), daemonSetName, metav1.GetOptions{})
 	if err != nil {
-		fmt.Println("Error ocurred for" + err.Error())
+		logrus.Infof("daemonset does not exist, err=%s", err.Error())
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply test/utils/daemonsets/daemonsets.go]: fixed

File: test/utils/l2discovery/l2discovery.go [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+}
+
+// Creates Mapping tables between interfaces index, mac address, and graph integer indexes
+func (config *L2DiscoveryConfig) createMaps(disc map[string]map[string]*Neighbors, nodeName string, index *int) {
+	for _, ifaceData := range disc[experimentalEthertype] {
+		config.ClusterMacToInt[ifaceData.Local.IfMac.Data] = *index
+		config.ClusterIndexToInt[IfClusterIndex{IfName: ifaceData.Local.IfName, NodeName: nodeName}] = *index
+		config.ClusterMacs[IfClusterIndex{IfName: ifaceData.Local.IfName, NodeName: nodeName}] = ifaceData.Local.IfMac.Data
+		config.ClusterIndexes[ifaceData.Local.IfMac.Data] = IfClusterIndex{IfName: ifaceData.Local.IfName, NodeName: nodeName}
+		aIf := PtpIf{}
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply test/utils/l2discovery/l2discovery.go]: fixed

File: test/utils/l2discovery/l2discovery.go [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+		return false
+	}
+
+	nic1, _ := getNicPort(ifaceName1.IfName)
+	nic2, _ := getNicPort(ifaceName2.IfName)
+	return nic1 == nic2
+}
+
+// gets the NIC port based on interface name (last numeral)
+func getNicPort(ifaceName string) (nic, port string) {
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply test/utils/l2discovery/l2discovery.go:332]: After discussions with Joseph, added a whitelist for ensXfY interfaces and added a TODO to use ethtool in the future

**Review by josephdrichard — APPROVED** [review]
/lgtm

**Review by josephdrichard — APPROVED** [review]
/lgtm

**Comment by aneeshkp** [comment]
@edcdavid need to rebase and resolve merge conflicts

**Pull request merged**